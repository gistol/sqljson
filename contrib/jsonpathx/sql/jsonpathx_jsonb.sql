CREATE EXTENSION jsonpathx;

-- map item method
select jsonb_path_query('1', 'strict $.map(x => x + 10)');
select jsonb_path_query('1', 'lax $.map(x => x + 10)');
select jsonb_path_query('[1, 2, 3]', '$.map(x => x + 10)');
select jsonb_path_query('[1, 2, 3]', '$.map(x => x + 10)[*]');
select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', '$.map(a => a.map(x => x + 10))');
select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', '$.flatmap(a => a.map(a => a + 10))');

-- map function
select jsonb_path_query('1', 'strict map($, x => x + 10)');
select jsonb_path_query('1', 'lax map($, x => x + 10)');
select jsonb_path_query('[1, 2, 3]', 'map($[*], x => x + 10)');
select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', 'map($[*], x => [map(x[*], x => x + 10)])');
select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', 'flatmap($[*], a => map(a[*], x => x + 10))');

-- reduce/fold item methods
select jsonb_path_query('1', 'strict $.reduce((x, y) => x + y)');
select jsonb_path_query('1', 'lax $.reduce((x, y) => x + y)');
select jsonb_path_query('1', 'strict $.fold((x, y) => x + y, 10)');
select jsonb_path_query('1', 'lax $.fold((x, y) => x + y, 10)');
select jsonb_path_query('[1, 2, 3]', '$.reduce((x, y) => x + y)');
select jsonb_path_query('[1, 2, 3]', '$.fold((x, y) => x + y, 100)');
select jsonb_path_query('[]', '$.reduce((x, y) => x + y)');
select jsonb_path_query('[]', '$.fold((x, y) => x + y, 100)');
select jsonb_path_query('[1]', '$.reduce((x, y) => x + y)');
select jsonb_path_query('[1, 2, 3]', '$.foldl((x, y) => [x, y], [])');
select jsonb_path_query('[1, 2, 3]', '$.foldr((x, y) => [y, x], [])');
select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', '$.fold((x, y) => x + y.fold((a, b) => a + b, 100), 1000)');

-- reduce/fold functions
select jsonb_path_query('1', 'strict reduce($, (x, y) => x + y)');
select jsonb_path_query('1', 'lax reduce($, (x, y) => x + y)');
select jsonb_path_query('1', 'strict fold($, (x, y) => x + y, 10)');
select jsonb_path_query('1', 'lax fold($, (x, y) => x + y, 10)');
select jsonb_path_query('[1, 2, 3]', 'reduce($[*], (x, y) => x + y)');
select jsonb_path_query('[1, 2, 3]', 'fold($[*], (x, y) => x + y, 100)');
select jsonb_path_query('[]', 'reduce($[*], (x, y) => x + y)');
select jsonb_path_query('[]', 'fold($[*], (x, y) => x + y, 100)');
select jsonb_path_query('[1]', 'reduce($[*], (x, y) => x + y)');
select jsonb_path_query('[1, 2, 3]', 'foldl($[*], (x, y) => [x, y], [])');
select jsonb_path_query('[1, 2, 3]', 'foldr($[*], (x, y) => [y, x], [])');
select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', 'fold($[*], (x, y) => x + y.fold((a, b) => a + b, 100), 1000)');

-- min/max item methods
select jsonb_path_query('1', 'strict $.min()');
select jsonb_path_query('1', 'lax $.min()');
select jsonb_path_query('[]', '$.min()');
select jsonb_path_query('[]', '$.max()');
select jsonb_path_query('[null]', '$.min()');
select jsonb_path_query('[null]', '$.max()');
select jsonb_path_query('[1, 2, 3]', '$.min()');
select jsonb_path_query('[1, 2, 3]', '$.max()');
select jsonb_path_query('[2, 3, 5, null, 1, 4, null]', '$.min()');
select jsonb_path_query('[2, 3, 5, null, 1, 4, null]', '$.max()');
select jsonb_path_query('["aa", null, "a", "bbb"]', '$.min()');
select jsonb_path_query('["aa", null, "a", "bbb"]', '$.max()');
select jsonb_path_query('[1, null, "2"]', '$.max()');

-- min/max functions
select jsonb_path_query('1', 'strict min($)');
select jsonb_path_query('1', 'lax min($)');
select jsonb_path_query('[]', 'min($[*])');
select jsonb_path_query('[]', 'max($[*])');
select jsonb_path_query('[null]', 'min($[*])');
select jsonb_path_query('[null]', 'max($[*])');
select jsonb_path_query('[1, 2, 3]', 'min($[*])');
select jsonb_path_query('[1, 2, 3]', 'max($[*])');
select jsonb_path_query('[2, 3, 5, null, 1, 4, null]', 'min($[*])');
select jsonb_path_query('[2, 3, 5, null, 1, 4, null]', 'max($[*])');
select jsonb_path_query('["aa", null, "a", "bbb"]', 'min($[*])');
select jsonb_path_query('["aa", null, "a", "bbb"]', 'max($[*])');
select jsonb_path_query('[1, null, "2"]', 'max($[*])');

-- tests for simplified variable-based lambda syntax
select jsonb_path_query('[1, 2, 3]', '$.map($1 + 100)');
select jsonb_path_query('[1, 2, 3]', 'map($[*], $1 + 100)');
select jsonb_path_query('[1, 2, 3]', '$.reduce($1 + $2)');
select jsonb_path_query('[1, 2, 3]', 'reduce($[*], $1 + $2)');
select jsonb_path_query('[1, 2, 3]', '$.fold($1 + $2, 100)');
select jsonb_path_query('[1, 2, 3]', 'fold($[*], $1 + $2, 100)');

-- more complex tests
select jsonb_path_query('[0,1,2,3,4,5,6,7,8,9]', '$.map((x,i,a) => {n: i, sum: reduce(a[0 to i], (x,y) => x + y)})[*]');
select jsonb_path_query('[0,1,2,3,4,5,6,7,8,9]', '$.fold((x,y,i,a) => [x[*], {n:y, s: [a[0 to i]].reduce($1+$2)}], [])[*]');
select jsonb_path_query('[0,1,2,3,4,5,6,7,8,9]', '$.fold((x,y) => [y,y,y].map((a) => a + y).reduce((x,y)=>x+y) + x * 100, 0)');

DROP EXTENSION jsonpathx;
