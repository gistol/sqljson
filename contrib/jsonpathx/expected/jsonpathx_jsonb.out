CREATE EXTENSION jsonpathx;
-- map item method
select jsonb_path_query('1', 'strict $.map(x => x + 10)');
ERROR:  SQL/JSON array not found
DETAIL:  jsonpath .map() is applied to not an array
select jsonb_path_query('1', 'lax $.map(x => x + 10)');
 jsonb_path_query 
------------------
 11
(1 row)

select jsonb_path_query('[1, 2, 3]', '$.map(x => x + 10)');
 jsonb_path_query 
------------------
 [11, 12, 13]
(1 row)

select jsonb_path_query('[1, 2, 3]', '$.map(x => x + 10)[*]');
 jsonb_path_query 
------------------
 11
 12
 13
(3 rows)

select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', '$.map(a => a.map(x => x + 10))');
            jsonb_path_query            
----------------------------------------
 [[11, 12], [13, 14, 15], [], [16, 17]]
(1 row)

select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', '$.flatmap(a => a.map(a => a + 10))');
       jsonb_path_query       
------------------------------
 [11, 12, 13, 14, 15, 16, 17]
(1 row)

-- map function
select jsonb_path_query('1', 'strict map($, x => x + 10)');
 jsonb_path_query 
------------------
 11
(1 row)

select jsonb_path_query('1', 'lax map($, x => x + 10)');
 jsonb_path_query 
------------------
 11
(1 row)

select jsonb_path_query('[1, 2, 3]', 'map($[*], x => x + 10)');
 jsonb_path_query 
------------------
 11
 12
 13
(3 rows)

select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', 'map($[*], x => [map(x[*], x => x + 10)])');
 jsonb_path_query 
------------------
 [11, 12]
 [13, 14, 15]
 []
 [16, 17]
(4 rows)

select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', 'flatmap($[*], a => map(a[*], x => x + 10))');
 jsonb_path_query 
------------------
 11
 12
 13
 14
 15
 16
 17
(7 rows)

-- reduce/fold item methods
select jsonb_path_query('1', 'strict $.reduce((x, y) => x + y)');
ERROR:  SQL/JSON array not found
DETAIL:  jsonpath .reduce() is applied to not an array
select jsonb_path_query('1', 'lax $.reduce((x, y) => x + y)');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('1', 'strict $.fold((x, y) => x + y, 10)');
ERROR:  SQL/JSON array not found
DETAIL:  jsonpath .fold() is applied to not an array
select jsonb_path_query('1', 'lax $.fold((x, y) => x + y, 10)');
 jsonb_path_query 
------------------
 11
(1 row)

select jsonb_path_query('[1, 2, 3]', '$.reduce((x, y) => x + y)');
 jsonb_path_query 
------------------
 6
(1 row)

select jsonb_path_query('[1, 2, 3]', '$.fold((x, y) => x + y, 100)');
 jsonb_path_query 
------------------
 106
(1 row)

select jsonb_path_query('[]', '$.reduce((x, y) => x + y)');
 jsonb_path_query 
------------------
(0 rows)

select jsonb_path_query('[]', '$.fold((x, y) => x + y, 100)');
 jsonb_path_query 
------------------
 100
(1 row)

select jsonb_path_query('[1]', '$.reduce((x, y) => x + y)');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('[1, 2, 3]', '$.foldl((x, y) => [x, y], [])');
 jsonb_path_query  
-------------------
 [[[[], 1], 2], 3]
(1 row)

select jsonb_path_query('[1, 2, 3]', '$.foldr((x, y) => [y, x], [])');
 jsonb_path_query  
-------------------
 [[[[], 3], 2], 1]
(1 row)

select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', '$.fold((x, y) => x + y.fold((a, b) => a + b, 100), 1000)');
 jsonb_path_query 
------------------
 1428
(1 row)

-- reduce/fold functions
select jsonb_path_query('1', 'strict reduce($, (x, y) => x + y)');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('1', 'lax reduce($, (x, y) => x + y)');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('1', 'strict fold($, (x, y) => x + y, 10)');
 jsonb_path_query 
------------------
 11
(1 row)

select jsonb_path_query('1', 'lax fold($, (x, y) => x + y, 10)');
 jsonb_path_query 
------------------
 11
(1 row)

select jsonb_path_query('[1, 2, 3]', 'reduce($[*], (x, y) => x + y)');
 jsonb_path_query 
------------------
 6
(1 row)

select jsonb_path_query('[1, 2, 3]', 'fold($[*], (x, y) => x + y, 100)');
 jsonb_path_query 
------------------
 106
(1 row)

select jsonb_path_query('[]', 'reduce($[*], (x, y) => x + y)');
 jsonb_path_query 
------------------
(0 rows)

select jsonb_path_query('[]', 'fold($[*], (x, y) => x + y, 100)');
 jsonb_path_query 
------------------
 100
(1 row)

select jsonb_path_query('[1]', 'reduce($[*], (x, y) => x + y)');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('[1, 2, 3]', 'foldl($[*], (x, y) => [x, y], [])');
 jsonb_path_query  
-------------------
 [[[[], 1], 2], 3]
(1 row)

select jsonb_path_query('[1, 2, 3]', 'foldr($[*], (x, y) => [y, x], [])');
 jsonb_path_query  
-------------------
 [[[[], 3], 2], 1]
(1 row)

select jsonb_path_query('[[1, 2], [3, 4, 5], [], [6, 7]]', 'fold($[*], (x, y) => x + y.fold((a, b) => a + b, 100), 1000)');
 jsonb_path_query 
------------------
 1428
(1 row)

-- min/max item methods
select jsonb_path_query('1', 'strict $.min()');
ERROR:  SQL/JSON array not found
DETAIL:  jsonpath .min() is applied to not an array
select jsonb_path_query('1', 'lax $.min()');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('[]', '$.min()');
 jsonb_path_query 
------------------
(0 rows)

select jsonb_path_query('[]', '$.max()');
 jsonb_path_query 
------------------
(0 rows)

select jsonb_path_query('[null]', '$.min()');
 jsonb_path_query 
------------------
 null
(1 row)

select jsonb_path_query('[null]', '$.max()');
 jsonb_path_query 
------------------
 null
(1 row)

select jsonb_path_query('[1, 2, 3]', '$.min()');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('[1, 2, 3]', '$.max()');
 jsonb_path_query 
------------------
 3
(1 row)

select jsonb_path_query('[2, 3, 5, null, 1, 4, null]', '$.min()');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('[2, 3, 5, null, 1, 4, null]', '$.max()');
 jsonb_path_query 
------------------
 5
(1 row)

select jsonb_path_query('["aa", null, "a", "bbb"]', '$.min()');
 jsonb_path_query 
------------------
 "a"
(1 row)

select jsonb_path_query('["aa", null, "a", "bbb"]', '$.max()');
 jsonb_path_query 
------------------
 "bbb"
(1 row)

select jsonb_path_query('[1, null, "2"]', '$.max()');
ERROR:  SQL/JSON scalar required
DETAIL:  boolean lambda expression in jsonpath .max() returned Unknown
-- min/max functions
select jsonb_path_query('1', 'strict min($)');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('1', 'lax min($)');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('[]', 'min($[*])');
 jsonb_path_query 
------------------
(0 rows)

select jsonb_path_query('[]', 'max($[*])');
 jsonb_path_query 
------------------
(0 rows)

select jsonb_path_query('[null]', 'min($[*])');
 jsonb_path_query 
------------------
 null
(1 row)

select jsonb_path_query('[null]', 'max($[*])');
 jsonb_path_query 
------------------
 null
(1 row)

select jsonb_path_query('[1, 2, 3]', 'min($[*])');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('[1, 2, 3]', 'max($[*])');
 jsonb_path_query 
------------------
 3
(1 row)

select jsonb_path_query('[2, 3, 5, null, 1, 4, null]', 'min($[*])');
 jsonb_path_query 
------------------
 1
(1 row)

select jsonb_path_query('[2, 3, 5, null, 1, 4, null]', 'max($[*])');
 jsonb_path_query 
------------------
 5
(1 row)

select jsonb_path_query('["aa", null, "a", "bbb"]', 'min($[*])');
 jsonb_path_query 
------------------
 "a"
(1 row)

select jsonb_path_query('["aa", null, "a", "bbb"]', 'max($[*])');
 jsonb_path_query 
------------------
 "bbb"
(1 row)

select jsonb_path_query('[1, null, "2"]', 'max($[*])');
ERROR:  SQL/JSON scalar required
DETAIL:  boolean lambda expression in jsonpath .max() returned Unknown
-- tests for simplified variable-based lambda syntax
select jsonb_path_query('[1, 2, 3]', '$.map($1 + 100)');
 jsonb_path_query 
------------------
 [101, 102, 103]
(1 row)

select jsonb_path_query('[1, 2, 3]', 'map($[*], $1 + 100)');
 jsonb_path_query 
------------------
 101
 102
 103
(3 rows)

select jsonb_path_query('[1, 2, 3]', '$.reduce($1 + $2)');
 jsonb_path_query 
------------------
 6
(1 row)

select jsonb_path_query('[1, 2, 3]', 'reduce($[*], $1 + $2)');
 jsonb_path_query 
------------------
 6
(1 row)

select jsonb_path_query('[1, 2, 3]', '$.fold($1 + $2, 100)');
 jsonb_path_query 
------------------
 106
(1 row)

select jsonb_path_query('[1, 2, 3]', 'fold($[*], $1 + $2, 100)');
 jsonb_path_query 
------------------
 106
(1 row)

-- more complex tests
select jsonb_path_query('[0,1,2,3,4,5,6,7,8,9]', '$.map((x,i,a) => {n: i, sum: reduce(a[0 to i], (x,y) => x + y)})[*]');
  jsonb_path_query   
---------------------
 {"n": 0, "sum": 0}
 {"n": 1, "sum": 1}
 {"n": 2, "sum": 3}
 {"n": 3, "sum": 6}
 {"n": 4, "sum": 10}
 {"n": 5, "sum": 15}
 {"n": 6, "sum": 21}
 {"n": 7, "sum": 28}
 {"n": 8, "sum": 36}
 {"n": 9, "sum": 45}
(10 rows)

select jsonb_path_query('[0,1,2,3,4,5,6,7,8,9]', '$.fold((x,y,i,a) => [x[*], {n:y, s: [a[0 to i]].reduce($1+$2)}], [])[*]');
 jsonb_path_query  
-------------------
 {"n": 0, "s": 0}
 {"n": 1, "s": 1}
 {"n": 2, "s": 3}
 {"n": 3, "s": 6}
 {"n": 4, "s": 10}
 {"n": 5, "s": 15}
 {"n": 6, "s": 21}
 {"n": 7, "s": 28}
 {"n": 8, "s": 36}
 {"n": 9, "s": 45}
(10 rows)

select jsonb_path_query('[0,1,2,3,4,5,6,7,8,9]', '$.fold((x,y) => [y,y,y].map((a) => a + y).reduce((x,y)=>x+y) + x * 100, 0)');
 jsonb_path_query  
-------------------
 61218243036424854
(1 row)

DROP EXTENSION jsonpathx;
